src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/TagList.cs

```diff
@@ -0,0 +1,553 @@
+                }
+
+                for (int i = _tagsCount; i > index; i--)
+                {
+                    _overflowTags[i] = _overflowTags[i - 1];
+                }
+                _overflowTags[index] = item;
+                _tagsCount++;
+                return;
+            }
+
+            Debug.Assert(_tagsCount < 8 && index < 7);
+
+            switch (index)
+            {
+                case 0: Tag8 = Tag7; Tag7 = Tag6; Tag6 = Tag5; Tag5 = Tag4; Tag4 = Tag3; Tag3 = Tag2; Tag2 = Tag1; Tag1 = item; break;
+                case 1: Tag8 = Tag7; Tag7 = Tag6; Tag6 = Tag5; Tag5 = Tag4; Tag4 = Tag3; Tag3 = Tag2; Tag2 = item; break;
+                case 2: Tag8 = Tag7; Tag7 = Tag6; Tag6 = Tag5; Tag5 = Tag4; Tag4 = Tag3; Tag3 = item; break;
+                case 3: Tag8 = Tag7; Tag7 = Tag6; Tag6 = Tag5; Tag5 = Tag4; Tag4 = item; break;
+                case 4: Tag8 = Tag7; Tag7 = Tag6; Tag6 = Tag5; Tag5 = item; break;
+                case 5: Tag8 = Tag7; Tag7 = Tag6; Tag6 = item; break;
+                case 6: Tag8 = Tag7; Tag7 = item; break;
+                default:
+                    Debug.Assert(false); // we shouldn't come here
+                    return;
+            }
+            _tagsCount++;
+        }
+
+        /// <summary>
+        /// Removes the element at the specified index of the <see cref="T:System.Diagnostics.TagList" />.
+        /// </summary>
+        /// <param name="index">The zero-based index of the element to remove.</param>
+        /// <exception cref="T:System.ArgumentOutOfRangeException"> <paramref name="index" /> index is less than 0 or <paramref name="index" /> is greater than <see cref="M:System.Diagnostics.TagList.Count" />.</exception>
+        public void RemoveAt(int index)
+        {
+            if ((uint)index >= (uint)_tagsCount)
+            {
+                throw new ArgumentOutOfRangeException(nameof(index));
+            }
+
+            if (_overflowTags is not null)
+            {
+                for (int i = index; i < _tagsCount - 1; i++)
+                {
+                    _overflowTags[i] = _overflowTags[i + 1];
+                }
+
+                _tagsCount--;
+                return;
+            }
+
+            Debug.Assert(_tagsCount <= 8 && index <= 7);
+
+            switch (index)
+            {
+                case 0: Tag1 = Tag2; goto case 1;
+                case 1: Tag2 = Tag3; goto case 2;
+                case 2: Tag3 = Tag4; goto case 3;
+                case 3: Tag4 = Tag5; goto case 4;
+                case 4: Tag5 = Tag6; goto case 5;
+                case 5: Tag6 = Tag7; goto case 6;
+                case 6: Tag7 = Tag8; break;
+                case 7: break;
+            }
+            _tagsCount--;
+        }
+
+        /// <summary>
+        /// Removes all elements from the <see cref="T:System.Diagnostics.TagList" />.
+        /// </summary>
+        public void Clear() => _tagsCount = 0;
```

Comment:
I am not seeing nulling the array pointer would be good if we already paid the cost of the memory allocation, why we need to repay it again after clearing the list? Please let me know if you think otherwise.

Reply:
I do not foresee people reusing these very often to justify keeping the alloc around. That being said, the allocation is fairly small. If you think the allocation can be reused by using clear to avoid allocation for many tag use, that seems reasonable to me.

Reply:
Think about when users want to clear the list? they will need to do that to reuse it for adding added items to it. Also, this list is defined as a struct, so it is expected any allocated memory lifetime will be short anyway. I am not seeing any concern keeping the allocated array when clearing the list.